File: app/src/main/java/com/baijunty/scanner/ScannerView.java
package com.baijunty.scanner;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Point;
import android.os.Build;
import android.util.AttributeSet;
import android.view.Display;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.lifecycle.Lifecycle;
import androidx.preference.PreferenceManager;

import com.google.zxing.client.android.CaptureActivityHandler;
import com.google.zxing.client.android.ViewfinderView;
import com.google.zxing.client.android.camera.CameraManager;

public class ScannerView extends FrameLayout {
    private OnScanResultFound resultFoundListener=null;
    private ScannerManager manager;
    protected int sound;
    private Point theScreenResolution =new Point();
    public void setResultFoundListener(OnScanResultFound resultFoundListener) {
        this.resultFoundListener = resultFoundListener;
    }

    public ScannerManager getManager(){
        if (manager==null){
            manager= new ScannerManager(this);
        }
        return manager;
    }

    CameraManager cameraManager;

    public CameraManager getCameraManager() {
        if (cameraManager==null){
            cameraManager=new CameraManager(getContext().getApplicationContext());
        }
        return cameraManager;
    }

    public void setAspect(float aspect) {
        ViewfinderView viewfinderView=findViewById(R.id.viewfinder_view);
        viewfinderView.setAspect(aspect);
    }

    public ScannerView(@NonNull Context context) {
        super(context);
        layoutScannerView(null);
    }

    public ScannerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        layoutScannerView(attrs);
    }

    public ScannerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        layoutScannerView(attrs);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public ScannerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        layoutScannerView(attrs);
    }

    public void bindLifecycle(Lifecycle lifecycle){
        PreferenceManager.setDefaultValues(getContext(), R.xml.preferences, false);
        ViewfinderView viewfinderView=findViewById(R.id.viewfinder_view);
        viewfinderView.setCameraManager(getCameraManager());
        lifecycle.addObserver(getManager());
        getManager().addObserve(result -> {
            if (resultFoundListener!=null&&resultFoundListener.onFounded(result)){
                restartPreview();
            }
        });
        postInvalidateDelayed(200);
    }

    private void layoutScannerView(AttributeSet attrs) {
        if (getChildCount() <= 0) {
            View.inflate(getContext(), R.layout.capture, this);
            if (attrs!=null){
                TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.ScannerView);
                float aspect=a.getFloat(R.styleable.ScannerView_aspect,1.0f);
                setAspect(aspect);
                sound=a.getResourceId(R.styleable.ScannerView_sound_source,R.raw.beep);
                a.recycle();
            }
        }
    }

    public void restartPreview(){
        CaptureActivityHandler handler=getManager().getHandler();
        if (handler!=null){
            handler.sendEmptyMessageDelayed(R.id.restart_preview, 2000);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        WindowManager manager = (WindowManager)getContext().getSystemService(Context.WINDOW_SERVICE);
        for (int i = 0; i < getChildCount(); i++) {
            View v=getChildAt(i);
            if (v.getId()==R.id.preview_view&&manager!=null){
                Display display = manager.getDefaultDisplay();
                display.getSize(theScreenResolution);
                v.measure(theScreenResolution.x |  (2 << 30), theScreenResolution.y |  (2 << 30));
                break;
            }
        }
    }

    public void displayError(String msg) {
        if (msg==null){
            msg=getContext().getString(R.string.msg_camera_framework_bug);
        }
        TextView t=findViewById(R.id.status_view);
        t.setText(msg);
    }
}

File: app/src/main/java/com/baijunty/scanner/ScannerView.kt
package com.baijunty.scanner

import android.content.Context
import android.graphics.Point
import android.os.Build
import android.preference.PreferenceManager
import android.util.AttributeSet
import android.view.View
import android.view.WindowManager
import android.widget.FrameLayout
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import com.google.zxing.client.android.ViewfinderView
import com.google.zxing.client.android.camera.CameraManager

class ScannerView : FrameLayout{
    var resultFoundListener:OnScanResultFound?=null
    private var aspect=1.0f
    private val manager:ScannerManager by lazy {
        ScannerManager(this)
    }
    val cameraManager: CameraManager by lazy {
        CameraManager(context.applicationContext)
    }

    constructor(context: Context) : super(context) {
        layoutScannerView(null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        layoutScannerView(attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        layoutScannerView(attrs)
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(
        context,
        attrs,
        defStyleAttr,
        defStyleRes
    ) {
        layoutScannerView(attrs)
    }

    fun bindLifecycle(lifecycle: Lifecycle){
        PreferenceManager.setDefaultValues(context, R.xml.preferences, false)
        val viewfinderView:ViewfinderView=findViewById(R.id.viewfinder_view)
        viewfinderView.setCameraManager(cameraManager)
        viewfinderView.setAspect(aspect)
        lifecycle.addObserver(manager)
        manager.addObserve(Observer{
            if (resultFoundListener?.onFounded(it)==true){
                restartPreview()
            }
        })
        postInvalidateDelayed(200)
    }

    private fun layoutScannerView(attrs: AttributeSet?) {
        if (childCount <= 0) {
            View.inflate(context, R.layout.capture, this)
            attrs?.let {
                val a = context.obtainStyledAttributes(attrs, R.styleable.ScannerView)
                aspect=a.getFloat(R.styleable.ScannerView_aspect,1.0f)
                a.recycle()
            }
        }
    }

    fun restartPreview(){
        manager.handler?.sendEmptyMessageDelayed(R.id.restart_preview, 2000)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val w=measuredWidth
        val h=measuredHeight
        val manager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = manager.defaultDisplay
        val theScreenResolution = Point()
        display.getSize(theScreenResolution)
        for (index in 0 until childCount){
            val v=getChildAt(index)
            if (v.id==R.id.preview_view){
                v.measure(theScreenResolution.x or  (2 shl 30), theScreenResolution.y or  (2 shl 30))
            }
        }
        setMeasuredDimension(w,h)
    }

    fun displayError(msg:String=context.getString(R.string.msg_camera_framework_bug)) {
        val t=findViewById<TextView>(R.id.status_view)
        t.text=msg
    }


}